### ISSUE TITLE
potential TensorFlow benchmark improvements

### ISSUE THREAD

**Issue by rryan** [soumith/convnet-benchmarks#91]
Hi -- thanks for the benchmarks!

I noticed that you do a sparse-to-dense conversion and use softmax_cross_entropy_with_logits. Have you tried eliding the sparse-to-dense conversion and using [sparse_softmax_cross_entropy_with_logits](https://www.tensorflow.org/versions/r0.7/api_docs/python/nn.html#sparse_softmax_cross_entropy_with_logits)? In my experience it the sparse_ version is faster.

Also, reduce_mean [does not have a GPU kernel](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/reduction_ops_mean.cc). reduce_sum with a division would prevent a GPU -> CPU -> GPU step when calculating the loss.

**Comment by soumith** [comment]
@rryan i dont think the sparse softmax will improve the overall benchmarks, the time contributed by a 1000-way softmax is very little. I'll def change reduce_mean to reduce_sum + division, if you think it helps.

**Comment by rryan** [comment]
@soumith -- I can't say how much it would affect this benchmark. On a much simpler model -- [the CIFAR-10 multi-GPU example](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py) -- replacing reduce_mean sped up my batches by 5% or so on a k20.
